### ISSUE TITLE
componentWillReceiveProps ifs hell

### ISSUE THREAD

**Issue by ASJvine** [kentcdodds/ama#378]
Hello @kentcdodds ,

I'm still pretty new on React development, but I've already work on 3 big project using React+Redux and I see a pattern that I dislike a lot:

```
componentWillReceiveProps(nextProps) {
        if (nextProps.params.type === TYPE_NEW_USER) {
            this.modalUsername = this.props.showPopup(<NewUsernamePopup onClose={::this.closeUsernamePopup} />, USERNAME_POPUP_ID, true);
        }
        if (this.state.kind !== nextProps.kind || this.state.filter !== nextProps.filter || this.state.hashtags !== nextProps.hashtags) {
            this.setState({
                results: [],
                loading: true,
                kind: nextProps.kind,
                filter: nextProps.filter,
                hashtags: nextProps.hashtags
            }, () => this.manageResults(nextProps.results, false));
        }else{
            this.manageResults(nextProps.results, true);
        }
        this.managePages(nextProps.paging);
    }
```

I would like to avoid the **ifs** inside the componentWillReceiveProps. How do you handle it? We've analysed another project using **Flux** and _callback registration_. It looks like:
```
componentWillMount() {
       EntityStore.on(EntityActions.ENTITIES_LOADED, this.getData.bind(this));
       EntityActions.entitiesLoaded();
   }
```

The first event is emitted by the component, but afterwards the store emits the event and the component updates. Additionally a single store keeps its state and do not duplicate async calls if it already has the content.  I personally like to avoid the **ifs**, but I do NOT want to lose Redux (its community and tools).

How would you add the current logic (ifs) inside the componentWillReceiveProps outside the component? I would like to handle the logic in a service layer and not inside the component.

I would definitely appreciate to read your opinion around this, because I've been struggling to find a solutions that fits.

Thanks a million from Barcelona

Regards,

Alexandre

**Comment by lukaleli** [comment]
Hey @ASJvine !
If I can put my two cents in the topic: I was struggling with similar a problem some time ago. To address the issue I wrote a HOC that wraps component and you can assign callbacks to specific state changes in your wrapped component. I published this component to npm and it's called `map-props-changes-to-callbacks`. You can find it here: [https://www.npmjs.com/package/map-props-changes-to-callbacks](https://www.npmjs.com/package/map-props-changes-to-callbacks). Perhaps you can find it helpful to clean your code up :)

**Comment by ajwhite** [comment]
I'm a huge fan of [recompose](https://github.com/acdlite/recompose), so I'm naturally drawn to `map-props-changes-to-callbacks` -- that looks very cool. It's nice and reactive too, great tip!

**Comment by ASJvine** [comment]
Nice approach @lukaleli !

However I was not looking for a HOC. Your solution moves the logic from **componentWillReceiveProps** to another part of the component (or its wrapper). I'd like to have some logic on the action/reducer layer and maybe a listener on the component that reacts to store changes.

**Comment by lukaleli** [comment]
@ASJvine  But the logic of changing state happens only in the reducers. Components listen to the store changes via `connect`. You still need to specify what changes you want to react to in your component. There's no way around it. Even if you use something like global PubSub bus that events are published to from reducers you still need to attach listeners on the component level.

**Comment by kentcdodds** [comment]
For the record: I'm not a huge fan of HOCs like those in recompose. Any codebase that uses recompose I've always found very hard to follow.

I'm not 100% certain I understand what you're looking for @ASJvine... If your code has some conditions in how it behaves, then you're probably going to need to codify that business logic somewhere...

When you say:

> I would like to handle the logic in a service layer and not inside the component.

By "service layer" are you talking about a JavaScript utilities module or some kind of server or something? That's a little ambiguous in your question I'm afraid.

Honestly your code doesn't seem all that complex and I'm afraid abstracting it out would only make things worse 😕 

Good luck!

**Comment by alexandreserrajaumot** [comment]
@kentcdodds by service level I've meant actions/reducers. I'll try to bring as much logic as possible to the action/reducers in order to avoid the conditionals inside the componentWillReceivProps.

Thanks @lukaleli and @kentcdodds for your tips!
